sdk/servicebus/service-bus/test/batchReceiver.spec.ts

```diff
@@ -1190,4 +1191,245 @@ describe("Batching - disconnects", function(): void {
     settledMessageCount.should.equal(2, "Unexpected number of settled messages.");
     refreshConnectionCalled.should.be.greaterThan(0, "refreshConnection was not called.");
   });
+
+  it("returns messages if drain is in progress (receiveAndDelete)", async function(): Promise<
+    void
+  > {
+    // Create the sender and receiver.
+    sbClient = getServiceBusClient();
+    const { receiverClient, senderClient } = await getSenderReceiverClients(
+      sbClient,
+      TestClientType.UnpartitionedQueue,
+      TestClientType.UnpartitionedQueue
+    );
+    const receiver = receiverClient.createReceiver(ReceiveMode.receiveAndDelete);
+    const sender = senderClient.createSender();
+
+    // The first time `receiveMessages` is called the receiver link is created.
+    // The `receiver_drained` handler is only added after the link is created,
+    // which is a non-blocking task.
+    await receiver.receiveMessages(1, 1);
+
+    if (!receiver["_context"].batchingReceiver!.isOpen()) {
+      throw new Error(`Unable to initialize receiver link.`);
+    }
+
+    // Send a message so we have something to receive.
+    await sender.send(TestMessage.getSample());
+
+    // Since the receiver has already been initialized,
+    // the `receiver_drained` handler is attached as soon
+    // as receiveMessages is invoked.
+    // We remove the `receiver_drained` timeout after `receiveMessages`
+    // does it's initial setup by wrapping it in a `setTimeout`.
+    // This triggers the `receiver_drained` handler removal on the next
+    // tick of the event loop; after the handler has been attached.
+    setTimeout(() => {
+      // remove `receiver_drained` event
+      receiver["_context"].batchingReceiver!["_receiver"]!.removeAllListeners(
+        ReceiverEvents.receiverDrained
+      );
+    }, 0);
+
+    // We want to simulate a disconnect once the batching receiver is draining.
+    // We can detect when the receiver enters a draining state when `addCredit` is
+    // called while `drain` is set to true.
+    let didRequestDrain = false;
+    const addCredit = receiver["_context"].batchingReceiver!["_receiver"]!.addCredit;
+    receiver["_context"].batchingReceiver!["_receiver"]!.addCredit = function(credits) {
+      didRequestDrain = true;
```

Comment:
Perhaps move this inside of the if on 1243?

Reply:
Yes, that's a good call. I missed moving that while I was refactoring this spy.